amr-wind/physics/MultiTemperature.cpp

```diff
@@ -0,0 +1,203 @@
+    const auto& geom_vec = m_sim.repo().mesh().Geom();
+    const int nlevels = m_sim.repo().num_active_levels();
+    for (int level = 0; level < nlevels; ++level) {
+        const auto& geom = geom_vec[level];
+        const auto& dx = geom.CellSizeArray();
+        const auto& prob_lo = geom.ProbLoArray();
+        auto& velocity = m_velocity(level);
+        auto& surf_temp = m_surf_temp(level);
+        auto& surf_temp_lower = m_surf_temp_lower(level);
+        auto& surf_temp_upper = m_surf_temp_upper(level);
+        // Copy Temperature to gpu
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_xloc.begin(), m_xloc.end(),
+            device_xloc.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_yloc.begin(), m_yloc.end(),
+            device_yloc.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, m_Tloc.begin(), m_Tloc.end(),
+            device_Tloc.begin());
+        const auto* xloc_ptr = device_xloc.data();
+        const auto* yloc_ptr = device_yloc.data();
+        const auto* Tloc_ptr = device_Tloc.data();
+        for (amrex::MFIter mfi(velocity); mfi.isValid(); ++mfi) {
+            const auto& vbx = mfi.validbox();
+            auto level_surf_temp = surf_temp.array(mfi);
+            auto level_surf_temp_lower = surf_temp_lower.array(mfi);
+            auto level_surf_temp_upper = surf_temp_upper.array(mfi);
+            const unsigned mesh_size = m_xloc.size();
+            const int index =
+                std::max(0, int(m_current_time / m_separation) - 1);
+            amrex::Print() << "Checking issues:" << m_Tloc.size() << "  "
+                           << mesh_size << "   " << index
+                           << index * (mesh_size - 1) << "  "
+                           << (index + 1) * (mesh_size - 1) << std::endl;
+            amrex::ParallelFor(
+                vbx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    // compute the source term
+                    const amrex::Real x = prob_lo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real y = prob_lo[1] + (j + 0.5) * dx[1];
+                    amrex::Real residual = 10000;
+                    amrex::Real locT1 = 300;
+                    amrex::Real locT2 = 300;
+                    for (unsigned ii = 0; ii < mesh_size; ++ii) {
+                        const amrex::Real radius = std::sqrt(
+                            std::pow(x - xloc_ptr[ii], 2) +
+                            std::pow(y - yloc_ptr[ii], 2));
+                        if (radius < residual) {
+                            residual = radius;
+                            locT1 = Tloc_ptr[index * (mesh_size - 1) + ii];
+                            locT2 =
+                                Tloc_ptr[(index + 1) * (mesh_size - 1) + ii];
+                        }
+                    }
+                    level_surf_temp(i, j, k, 0) = locT1;
```

Comment:
would it be better to store these 3 numbers into a 3 component field instead of 3 different fields? It would remove a whole bunch of LOCs